Easy-to-use database migrations at application level

bajtos opened the pull request:
I find the current instructions for database migration as too involved for practical use, for example when running our example applications against a real database (SQL, MongoDB). I would like to propose a better user experience as a short-term solution to scratch this itch. Please see https://github.com/strongloop/loopback-next/issues/487 for the discussion about a more comprehensive solution for long term.

In this pull request, I am introducing new APIs to simplify database migrations:

- ~~A new interface `MigrateableRepository` describing repositores that know how to migrate their database schema.~~

- A new Application-level method `app.migrateSchema()` provided by `RepositoryMixin` and running schema migration provided by all registered ~~repositories~~ datasources under the hood.

Using these two new APIs, I also

- ~~Implement `MigrateableRepository` in `DefaultCrudRepository` using autoupdate/automigrate APIs provided by legacy-juggler's  DataSource class.~~

- Simplify the instructions shown in `Database-migrations.db`

- Add an example `migrate.ts` script to our Todo example to verify that the code snippet shown in the docs works as intended.

<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #123
Implements #254
See also #23
-->

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- out of scope: Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

raymondfeng commented on packages/repository/src/mixins/repository.mixin.ts:
```diff
@@ -163,6 +170,34 @@ export function RepositoryMixin<T extends Class<any>>(superClass: T) {
+     * @param options Migration options, e.g. whether to update tables
+     * preserving data or rebuild everything from scratch.
+     */
+    async migrateSchema(options?: SchemaMigrationOptions): Promise<void> {
```
I would prefer to have two methods and use `updateSchema` as default.

- migrateSchema(...)
- updateSchema(...)

raymondfeng replied:
Or at least an array of repo names.

raymondfeng commented on packages/repository/src/repositories/legacy-juggler-bridge.ts:
```diff
@@ -332,4 +336,9 @@ export class DefaultCrudRepository<T extends Entity, ID>
     return models.map(m => this.toEntity(m));
   }
+
+  async migrateSchema(options?: SchemaMigrationOptions): Promise<void> {
```
Same here - `automigrate` and `autoupdate`.  At repository level, maybe it's better to keep the names as `automigrate` and `autoupdate`.

raymondfeng commented on packages/repository/src/repositories/migrateable.repository.ts:
```diff
@@ -0,0 +1,49 @@
+   * When set to true, schema migration will drop existing tables and recreate
+   * them from scratch, removing any existing data along the way.
+   */
+  rebuild?: boolean;
```
We should have an option to dry run - returning a bunch of SQL statements.

bajtos replied:
+1

Unfortunately, our current juggler API does not provide dry-run feature, therefore I don't have any easy means for enabling dry run option. Let's leave it out of scope of this pull request.

In the future, I would like us to implement a mode where we create a SQL file containing the migration script, so that the user can customize it. See https://github.com/strongloop/loopback-next/issues/487.

raymondfeng commented on packages/repository/src/repositories/migrateable.repository.ts:
```diff
@@ -0,0 +1,49 @@
+   * When set to true, schema migration will drop existing tables and recreate
+   * them from scratch, removing any existing data along the way.
+   */
+  rebuild?: boolean;
```
The name `rebuild` is not clear. Maybe something like `dropThenRecreate`?

bajtos commented:
@raymondfeng 

> I would prefer to have two methods and use updateSchema as default.
>
> - `migrateSchema(...)`
> - `updateSchema(...)`

Let's discuss.

I personally prefer a single method with an option to choose between an incremental update or destructive drop + create. I makes it easier to implement wrapper functions delegating the implementation to lower-level layers, e.g. a single `app.migrateSchema` calling a single `repository.migrateSchema` & forwarding the entire `options` argument.

When there are two methods, then we also need multiple wrapper functions (`app.migrateSchema` and `app.updateSchema`.

What are your arguments for having two methods? 

I also find the names "update" and "migrate" not descriptive enough, the term "migrate" does not imply drop+create in my mind. AFAIK, the term "migration" is usually used for non-destructive upgrades. Few references:
 - Ruby on Rails https://guides.rubyonrails.org/v3.2/migrations.html.
 - Martin Fowler on Evolutionary Database Design: https://www.martinfowler.com/articles/evodb.html#AllDatabaseChangesAreMigrations
 - .NET EntityFramework's Code First Migrations: https://docs.microsoft.com/en-us/aspnet/mvc/overview/getting-started/getting-started-with-ef-using-mvc/migrations-and-deployment-with-the-entity-framework-in-an-asp-net-mvc-application

> It should be possible to filter what repos should be migrated/updated.

I don't understand what's the use case for such filter.

Here is the user story I am trying to address: _As a developer wanting to run a LB4 app, I want an easy way how to update my database schema to match application's domain model._  In this scenario, we always want to update all tables.

What scenario do you @raymondfeng have in mind for choosing a subset of tables only? I think this filter is easy to add later as an incremental improvement of my MVP solution, thus I am proposing to leave it out of this initial pull request.

bajtos commented:
In d8c52a6, I reworked the underlying implementation to leverage DataSource-level `autoupdate`/`automigrate` API and thus allow connectors to set up foreign keys.

The pull request is ready for another round of review. @strongloop/loopback-maintainers PTAL.

I tested the migration with `examples/todo-list` and MySQL connector. The tables were created correctly, but unfortunately the foreign key constraint was not set up. I don't know whether this is a limitation of our MySQL connector, or whether it's a problem of the relation implementation in LB4. Either way, I consider this issue out of scope of this (initial) pull request.

raymondfeng commented on packages/repository/src/datasource.ts:
```diff
@@ -17,3 +17,11 @@ export interface DataSource {
+   * When set to true, schema migration will drop existing tables and recreate
+   * them from scratch, removing any existing data along the way.
+   */
+  dropExistingSchema?: boolean;
```
Maybe instead of `boolean`, we should have an `enum`:

```ts
policyForExistingSchema: 'drop' | 'alter' | 'ignore' | 'fail';
```

bajtos replied:
I am not sure what `ignore` would mean, but other than that I like the idea to use an enum that's easier to extend in the future.

The current juggler implementation supports `drop` and `alter` only, so that's what I am going to use in this pull request.

I find `policyForExistingSchema` too long, how about shorter `existingSchema`?

raymondfeng replied:
We can start with `'drop' | 'alter'`. I'm fine with `existingSchema`.

raymondfeng commented:
> The tables were created correctly, but unfortunately the foreign key constraint was not set up. I don't know whether this is a limitation of our MySQL connector, or whether it's a problem of the relation implementation in LB4.

I think it's a limitation of LB3 `automigrate/autoupdate`.

bajtos commented:
> I think it's a limitation of LB3 automigrate/autoupdate.

Good to know, so I am not going to worry about this for now. I think the most important thing is that the current high-level design does not prevent the connector to setup referential integrity and possibly other constraints in the future.

raymondfeng commented:
> What scenario do you @raymondfeng have in mind for choosing a subset of tables only? I think this filter is easy to add later as an incremental improvement of my MVP solution, thus I am proposing to leave it out of this initial pull request.

Developers want to control what models that need to be migrated. For example, our UI allows people to select a list of models and click `migrate`.

In real world, some models are created newly while others are discovered from existing DB schema. We don't want run `migrate` for such models discovered.

We can simply add `models` to the `options` and pass it down to datasource.

bajtos commented:
> Developers want to control what models that need to be migrated. For example, our UI allows people to select a list of models and click migrate.
> 
> In real world, some models are created newly while others are discovered from existing DB schema. We don't want run migrate for such models discovered.
> 
> We can simply add models to the options and pass it down to datasource.

I feel it's a bit of a scope creep here, but I am ok to add models to the options in the TS/JS API as you suggested :+1:

bajtos commented:
@raymondfeng I added support for `options.models` (defaulting to migrate all models) and reworked a boolean option `dropExistingSchema` into `existingSchema` accepting a string enum.

LGTY now?

---

I would also like to hear opinions from @strongloop/loopback-maintainers on the following topic I mentioned earlier:

Right now, the pull request contains what I consider as a minimal viable implementation/improvement. Besides the changes proposed, I would like to discuss few more improvements we may or may not want to do, either as part of this pull request or in follow-ups:

- Add `src/migrate.ts` to all example repositories. I think it would be very useful to have automigration available in our TodoList example too. /cc @b-admike 
- Introduce an npm/package.json script, e.g. `npm run update-schema`, as an alias for `node dist/src/migrate`.
- Modify the CLI template to include `src/migrate.ts` and `npm run update-schema` in all new projects.

dhmlau commented:
IMHO, it is a good idea to have a more "automatic" way to create the database tables for users, especially for new users.    

Another question or discussion point is that what we want for the default behaviour for someone creating a LB app for real use.  Maybe it's just me, I might just want the `migrate` script to be run for the first time to create the db definitions or whenever that's needed rather than on every application start.

b-admike commented on docs/site/Database-migrations.md:
````diff
@@ -89,6 +93,35 @@ your database by running `node dist/src/migrate` and rebuild it from scratch by
+{% include code-caption.html content="src/application.ts" %}
+
+```ts
+export class TodoListApplication extends BootMixin(
````
nitpick: maybe add a comment for imports on top?

b-admike commented on examples/todo/src/migrate.ts:
```diff
@@ -6,12 +6,20 @@
+  await app.migrateSchema({dropExistingSchema});
+
+  // Connectors usually keep a pool of opened connections,
+  // this keeps the process running even after all works is done.
```
nitpick: `works` -> `work`

b-admike commented on packages/repository/src/mixins/repository.mixin.ts:
```diff
@@ -184,17 +180,30 @@ export function RepositoryMixin<T extends Class<any>>(superClass: T) {
+      const repoBindings: Readonly<Binding<unknown>>[] = this.findByTag(
+        'repository',
+      );
+      await Promise.all(repoBindings.map(b => this.get(b.key)));
```
I'm wondering how this will affect performance if we're loading up all the repository instances? Should we do some sort of filtering now that we support select models in the `migrateSchema` function?

bajtos replied:
Great question!

We are creating instances of repository classes. This is a lightweight task - a repository instance is created and a new PersistedModel class is registered with the appropriate datasource. I am not anticipating any performance issues. 

Migration is a one-time action so even if its performance is not optimal, the affect on user applications should be minimal.

To allow datasource migration to define foreign key constraints, it may be necessary to setup related models even if they are not migrated. For example, if we are migrating only an Order model which has a belongs to relation to User, then we should instantiate UserRepository too, so that the datasource/connector knows about both sides of the relation.

Finally, I am not aware of a reasonably easy & sufficiently robust way of how to filter repositories based on model names, because the repository binding does not provide any information about the model it is providing data access to.

b-admike replied:
Thank you for clarifying 💯!
